### PR TITLE
Support caching of Kerberos Subject

### PR DESCRIPTION
**Pull request by jyemin**
* Add SubjectProvider interface
* Add JAVA_SUBJECT_PROVIDER mechanism property
* Add KerberosSubjectProvider implementation
* Default to use KerberosSubjectProvider with "com.sun.security.jgss.krb5.initiate" LoginContext name

JAVA-3836

### CODE REVIEW

File: .evergreen/run-gssapi-auth-test.sh [hunk 1]
```diff
@@ -41,4 +41,4 @@ echo "Running tests with ${JDK}"
 -Dorg.mongodb.test.uri=${MONGODB_URI} \
 -Pgssapi.enabled=true -Psun.security.krb5.debug=true -Pauth.login.config=file://${PROJECT_DIRECTORY}/.evergreen/java.login.drivers.config \
 -Pkrb5.kdc=${KDC} -Pkrb5.realm=${REALM} -Psun.security.krb5.debug=true \
-driver-core:test --tests GSSAPIAuthenticationSpecification
+driver-core:test --tests GSSAPIAuthenticationSpecification --tests GSSAPIAuthenticatorSpecification --tests KerberosSubjectProviderTest
```
Comment (jyemin) [inline .evergreen/run-gssapi-auth-test.sh:44]: Almost forgot to do this...  But this is the only way to get the new tests to run

File: driver-core/src/main/com/mongodb/internal/connection/GSSAPIAuthenticator.java [hunk 2]
```diff
@@ -98,4 +101,9 @@ private String getHostName(final ServerAddress serverAddress) throws UnknownHost
                ? InetAddress.getByName(serverAddress.getHost()).getCanonicalHostName()
                : serverAddress.getHost();
     }
+
+    @NonNull
+    protected SubjectProvider getDefaultSubjectProvider() {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/internal/connection/GSSAPIAuthenticator.java:106]: This is how we get the default per-MongoClient caching behavior

File: driver-core/src/main/com/mongodb/KerberosSubjectProvider.java [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            loginContext.login();
+            subject = loginContext.getSubject();
+            LOGGER.info("Login successful");
+        }
+        return subject;
+    }
+
+    private static boolean needNewSubject(@NonNull final Subject subject) {
+        for (KerberosTicket cur : subject.getPrivateCredentials(KerberosTicket.class)) {
+            if (cur.getServer().getName().startsWith("krbtgt/")) {
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/KerberosSubjectProvider.java:93]: Nit: Make the "krbtgt/" string static.

File: driver-core/src/test/functional/com/mongodb/internal/connection/GSSAPIAuthenticationSpecification.groovy [hunk 4]
```diff
@@ -37,13 +38,13 @@ import javax.security.auth.login.LoginContext
...
 import static com.mongodb.ClusterFixture.getConnectionString
-import static com.mongodb.ClusterFixture.getCredential
 import static com.mongodb.ClusterFixture.getSslSettings
+import static com.mongodb.MongoCredential.JAVA_SUBJECT_PROVIDER_KEY
 import static com.mongodb.MongoCredential.createGSSAPICredential
 import static com.mongodb.internal.connection.CommandHelper.executeCommand
 import static java.util.concurrent.TimeUnit.SECONDS
 
-@IgnoreIf({ getCredential() == null || getCredential().getAuthenticationMechanism() != GSSAPI })
+@IgnoreIf({ ClusterFixture.getCredential() == null || ClusterFixture.getCredential().getAuthenticationMechanism() != GSSAPI })
```
Comment (rozza) [inline driver-core/src/test/functional/com/mongodb/internal/connection/GSSAPIAuthenticationSpecification.groovy:47]: Nit: Mixed import static methods from ClusterFixture - some `getConnectionString` & `getSslSettings` used directly. `getCredential()` is no longer - was that on purpose?

**Review by rozza — COMMENTED** [review]
Some minor nits - but LGTM after they're cleaned up

File: driver-core/src/test/functional/com/mongodb/internal/connection/GSSAPIAuthenticationSpecification.groovy (hunk 4 above)
Comment (jyemin) [reply driver-core/src/test/functional/com/mongodb/internal/connection/GSSAPIAuthenticationSpecification.groovy]: Reverted.

File: driver-core/src/main/com/mongodb/KerberosSubjectProvider.java (hunk 3 above)
Comment (jyemin) [reply driver-core/src/main/com/mongodb/KerberosSubjectProvider.java]: Done

**Review by rozza — APPROVED** [review]
LGTM!

**Review by durran — APPROVED** [review]
LGTM

**Pull request merged**